lib/src/rules/prefer_is_empty.dart

```diff
@@ -0,0 +1,210 @@
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:analyzer/src/dart/constant/evaluation.dart';
+import 'package:analyzer/src/dart/constant/value.dart';
+import 'package:analyzer/src/error/codes.dart';
+import 'package:analyzer/src/generated/engine.dart';
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:linter/src/linter.dart';
+
+const desc = 'Use isNotEmpty for Iterables and Maps.';
+
+const useIsNotEmpty = 'Use isNotEmpty instead of length';
+const useIsEmpty = 'Use isEmpty instead of length';
+const alwaysFalse = 'Always false because length is always greater or equal 0.';
+const alwaysTrue = 'Always true because length is always greater or equal 0.';
+
+const details = '''
+**DO NOT** use `.length` to see if a collection is empty.
+
+The `Iterable` contract does not require that a collection know its length or be
+able to provide it in constant time. Calling `.length` just to see if the
+collection contains anything can be painfully slow.
+
+Instead, there are faster and more readable getters: `.isEmpty` and
+`.isNotEmpty`. Use the one that doesn’t require you to negate the result.
+
+**GOOD:**
+```
+if (lunchBox.isEmpty) return 'so hungry...';
+if (words.isNotEmpty) return words.join(' ');
+```
+
+**BAD:**
+```
+if (lunchBox.length == 0) return 'so hungry...';
+if (words.length != 0) return words.join(' ');
+```
+''';
+
+class _LintCode extends LintCode {
+  static final registry = <String, LintCode>{};
+
+  factory _LintCode(String name, String message) => registry.putIfAbsent(
+      name + message, () => new _LintCode._(name, message));
+
+  _LintCode._(String name, String message) : super(name, message);
+}
+
+class PreferIsEmpty extends LintRule {
+  PreferIsEmpty()
+      : super(
+            name: 'prefer_is_empty',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+
+  void reportLintWithDescription(AstNode node, String description) {
+    if (node != null) {
+      reporter.reportErrorForNode(new _LintCode(name, description), node, []);
+    }
+  }
+}
+
+class Visitor extends SimpleAstVisitor {
+  final PreferIsEmpty rule;
+  Visitor(this.rule);
+
+  @override
+  visitSimpleIdentifier(SimpleIdentifier identifier) {
+    AstNode lengthAccess;
+    SimpleIdentifier lengthIdentifier;
+    InterfaceType type;
+
+    AstNode parent = identifier.parent;
+    if (parent is PropertyAccess) {
+      lengthIdentifier = parent.propertyName;
+      lengthAccess = parent;
+      if (parent.target?.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.target?.bestType;
+    } else if (parent is PrefixedIdentifier) {
+      lengthIdentifier = parent.identifier;
+      lengthAccess = parent;
+      if (parent.prefix.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.prefix.bestType;
+    }
```

Comment:
I might be missing something, but it looks like this would generate two lints for every violation (though I'm somewhat surprised if the test framework doesn't catch that issue). Given an expression like `list.length`, this method will be called twice: one for `list` and once for `length`, and both times we will exit this `if` with a non-null `lengthIdentifier`.

I think you need to do something like
```
if (parent is PropertyAccess && identifier == parent.propertyName) {
```
In which case you probably don't need `lengthIdentifier` any more. And if you add an `} else { return;` to the `if`, you won't need `lengthAccess` either. (I'm not requesting either of these last two changes, just noting them.)